### PR TITLE
Bugfix/axiom curry

### PR DESCRIPTION
**Pull request by leissa**
For a curried App of an Axiom, you only want to trigger normalization at specific spots. For this reason, Thorin maintains a Def::curry_ counter that each App decrements. The Axiom::normalizer() will be triggered when Axiom::curry() becomes `0`. These are also the spots that you can match/force. After that, the counter will be set to Axiom::trip(). E.g., let's say an Axiom has this type:
```
A -> B -> C -> D -> E
          ^         |
          |         |
          +---------+
```
Using an initial value as `5` for Axiom::curry and `3` as Axiom::trip has the effect that here
```
x a b c1 d1 e1 c2 d2 e2 c3 d3 e3
```
the Axiom::normalizer will be triggered after App'ing `e1`, `e2`, and `e3`

In the above case you can use Axiom::infer_curry_and_trip to automatically retrieve these values (happens as default in the frontend). You can specify other values, as you like in the frontend after the normalizer name:
```
.ax foo: A -> B -> C -> D -> E, normalize_foo, 3;    // curry 3, trip 0
.ax foo: A -> B -> C -> D -> E, normalize_foo, 3, 1; // curry 3, trip 1
```
See also the unit test `Axiom::curry` for details.

### CODE REVIEW

File: dialects/autodiff/auxiliary/autodiff_aux.cpp [hunk 1]
```diff
@@ -146,7 +146,7 @@ const Def* zero_def(const Def* T) {
 const Def* op_sum(const Def* T, DefArray defs) {
     // TODO: assert all are of type T
     auto& world = T->world();
-    return world.raw_app(world.raw_app(world.ax<sum>(), {world.lit_nat(defs.size()), T}), world.tuple(defs));
+    return world.app<false>(world.app<false>(world.ax<sum>(), {world.lit_nat(defs.size()), T}), world.tuple(defs));
```
Comment (NeuralCoder3) [inline dialects/autodiff/auxiliary/autodiff_aux.cpp:149]: We want to normalize at this point.
The sum should be resolved to a chain of additions.

I think that most of the following `<false>` should be `<true>` in the general spirit of "always apply normalizers" (except the normalizer pass-through cases).

Comment (leissa) [reply dialects/autodiff/auxiliary/autodiff_aux.cpp:149]: So, you are telling me, the original code should have used `app` instead of `raw_app`?

Comment (NeuralCoder3) [reply dialects/autodiff/auxiliary/autodiff_aux.cpp:149]: Probably. I think I misunderstood `raw_app` and some of my code is broken/does not work as intended.

**Comment by leissa** [comment]
@NeuralCoder3 I re-introduced the `raw_app` (now receiving the `type` of the `App`) because we don't need to recompute the dependent type twice. I also changed a couple of `raw_app`s in autodiff with proper `app`s. Seems to work.

I've also merged your latest work with inline optimizations into master and now in to this branch. Can you have a look whether this magic here actually fixes your problems?

**Comment by NeuralCoder3** [comment]
> I've also merged your latest work with inline optimizations into master and now in to this branch. Can you have a look whether this magic here actually fixes your problems?

Yes, the normalization for recursive types and higher-order returns are now working.

File: dialects/direct/direct.thorin [hunk 2]
```diff
@@ -22,7 +22,7 @@
 /// This axiom lets the user call a cps function in direct style.
 /// The function is not converted. Only the call site is changed.
 /// 
-.ax %direct.cps2ds: Π [T: *, U: *] -> (.Cn [T, .Cn U]) -> (T -> U), normalize_cps2ds;
+.ax %direct.cps2ds: Π [T: *, U: *] -> (.Cn [T, .Cn U]) -> (T -> U), normalize_cps2ds, 2;
```
Comment (leissa) [inline dialects/direct/direct.thorin:25]: Nice :)

**Pull request merged**